website/src/docs/hotchocolate/server/instrumentation.md

```diff
@@ -0,0 +1,278 @@
+
+```csharp
+public override IDisposable ExecuteRequest(IRequestContext context)
+{
+    _logger.LogInformation("Request execution started!");
+
+    return EmptyScope;
+}
+```
+
+# Execution Events
+
+We can hook into execution events of the Hot Chocolate execution engine by creating a class inheriting from `ExecutionDiagnosticEventListener`.
+
+```csharp
+public class MyExecutionEventListener : ExecutionDiagnosticEventListener
+{
+    public override IDisposable ExecuteRequest(IRequestContext context)
+    {
+        // Omitted code for brevity
+    }
+}
+```
+
+The following methods can be overriden.
+
+| Method name                  | Description                                                                                                                                    |
+| ---------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------- |
+| ExecuteRequest               | Scope that encloses the entire GraphQL request execution. Also the first diagnostic event raised during a GraphQL request.                     |
+| RequestError                 | Called if the GraphQL request produced an error. Called immediately before the scope of `ExecuteRequest` is disposed.                          |
+| ExecuteSubscription          | Scope that encloses the execution of a subscription query. Scope is created once a client subscribes and disposed once the subscription ends.  |
+| ParseDocument                | Scope that encloses the parsing of a document.                                                                                                 |
+| SyntaxError                  | Called if a document could not be parsed due to a syntax error.                                                                                |
+| ValidateDocument             | Scope that encloses the validation of a document.                                                                                              |
+| ValidationErrors             | Called if errors occured during the validation of the document.                                                                                |
+| StartProcessing              | TODO scope                                                                                                                                     |
+| StopProcessing               | TODO                                                                                                                                           |
+| RunTask                      | TODO scope                                                                                                                                     |
+| TaskError                    | TODO                                                                                                                                           |
```

Comment:
@michaelstaib I'm having a hard time grasping when these would be executed and what they represent. 
I always thought of Tasks as resolvers that need to be executed. But after looking at the code it seems that the `RunTask` event is only raised for the abstract `ExecutionTask`, but not a `ResolverTask` for example. 
I'm also having trouble with the `WorkScheduler` i.e. `Start/StopProcessing`.
I tried looking at the code, but it's a bit much 😅

Would be great if you could give me a quick summary sentence for when `StartProcessing`, `StopProcessing`, `RunTask` and `TaskError` are invoked. 

Bonus points if you could explain the `WorkScheduler` and `ExecutionTasks` on a high level for me! 😊

Reply:
The new execution engine works differently than 11. We essentially execute tasks of an execution plan. You can also bring your own special tasks that do something in the execution. The new stitching engine is the first thing really using that. Essentially we will create tasks that fetch in one go the data of multiple fields/resolvers and the push the completed result into the actual resolver pipeline.

For resolvers we have a specific event, for any other task we use RunTask.

Further StartProcessing/StopProcessing is invoked when the execution engine starts doing something like starting tasks, or invoking DataLoader and so on. Stop is invoked when the execution engine needs to wait for resolvers to complete to do more work or whenever the execution has completed.

The events can be used to inspect efficiency in the request execution.

Reply:
Awesome, thanks for the explanation! :)